examples/cfq/utils.py

```diff
@@ -0,0 +1,234 @@
+import tensorflow.compat.v2 as tf
+import tensorflow_datasets as tfds
+from absl import logging
+
+import constants
+
+
+def _get_tokens(input_features: Dict, 
+                tokenizer: text.Tokenizer,
+                datasets: Iterable[tf.data.Dataset],
+                preprocessing_fun: Any) -> Iterable[List[bytes]]:
+    """Returns a list of tokens for all input fields in the given datasets."""
+
+    def _tokenize_input_features(
+            example: Dict[Text, tf.Tensor]) -> Dict[Text, tf.Tensor]:
+        """Tokenizes all input features in an example."""
+        example = preprocessing_fun(example)
+        for feature in example:
+            if feature in input_features:
+                example[feature] = tokenizer.tokenize(example[feature])
+        return example
+
+    for dataset in datasets:
+        # Apply the tokenizer to all input features.
+        tokenized_dataset = dataset.map(_tokenize_input_features,
+                                        num_parallel_calls=constants.AUTOTUNE)
+        # Yield all tokenized input features (i.e., tokenized input sentences).
+        for example in tfds.as_numpy(tokenized_dataset):
+            for feature in input_features:
+                yield example[feature]
+
+
+def build_vocabulary(file_name: Text,
+                     input_features: Dict,
+                     tokenizer: text.Tokenizer,
+                     datasets: Iterable[tf.data.Dataset],
+                     special_tokens: Sequence[bytes] = (constants.PAD,
+                                                        constants.UNK,
+                                                        constants.BOS,
+                                                        constants.EOS),
+                     preprocessing_fun: Any = lambda x: x) -> Dict[bytes, int]:
+    """Returns a vocabulary of tokens with optional minimum frequency.
+    The vocabulary is saved to a temps file
+
+    Args:
+        
+        special_tokens: Special tokens that will be the start of the vocabulary.
+        tokenizer: tf tokenizer (eg. WhitespaceTokenizer)
+        datasets: datasets from which the vocabulary should be extracted
+        special_tokens: special tokens to be added in the vocabulary; they are
+                        added at the begginig with PAD at index 0
+        preprocessing_fun: function for applying preprocessing on a dataset
+                           entry; by default the identity function
+                           (no preprocessing)
+    Returns:
+        An ordered dictionary that maps tokens to their IDs in the vocabulary.
+    """
+
+    vocab_dir = 'temp'
+    vocab_path = vocab_dir+'/'+file_name
+    if not os.path.exists(vocab_dir):
+        os.makedirs(vocab_dir)
+    # Try to read the vocabulary
+    if os.path.isfile(vocab_path):
+        with open(vocab_path, 'rb') as vocab_file:
+            vocab = pickle.load(vocab_file)
+            vocab_file.close()
+            return vocab
```

Comment:
Let the context manager handle closing the file.

```suggestion
        with open(vocab_path, 'rb') as vocab_file:
            vocab = pickle.load(vocab_file)
        return vocab
```